cpp/src/link_analysis/pagerank.cu

```diff
@@ -1,286 +1,278 @@
+ * and any modifications thereto.  Any use, reproduction, disclosure or
+ * distribution of this software and related documentation without an express
+ * license agreement from NVIDIA CORPORATION is strictly prohibited.
+ *
+ */
+
+// Pagerank solver
+// Author: Alex Fender afender@nvidia.com
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stddef.h>
+#include <string>
+#include <sstream>
+#include <iostream>
+#include <iomanip>
+#include "cub/cub.cuh"
+#include <algorithm>
+#include <iomanip>
+
+#include <rmm_utils.h>
+
+#include "utilities/graph_utils.cuh"
+#include "utilities/error_utils.h"
+#include <cugraph.h>
+#include <graph.hpp>
+
+namespace cugraph { 
+namespace detail {
+
+#ifdef DEBUG
+  #define PR_VERBOSE
+#endif
+
+template <typename IndexType, typename ValueType>
+bool pagerankIteration(IndexType n, IndexType e, IndexType const *cscPtr, IndexType const *cscInd,ValueType *cscVal,
+                       ValueType alpha, ValueType *a, ValueType *b, float tolerance, int iter, int max_iter,
+                       ValueType * &tmp,  void* cub_d_temp_storage, size_t  cub_temp_storage_bytes,
+                       ValueType * &pr, ValueType *residual) {
+    ValueType  dot_res;
+    CUDA_TRY(cub::DeviceSpmv::CsrMV(cub_d_temp_storage, cub_temp_storage_bytes, cscVal,
+                                    (IndexType *) cscPtr, (IndexType *) cscInd, tmp, pr, n, n, e));
+
+    scal(n, alpha, pr);
+    dot_res = dot( n, a, tmp);
+    axpy(n, dot_res,  b,  pr);
+    scal(n, (ValueType)1.0/nrm2(n, pr) , pr);
+    axpy(n, (ValueType)-1.0,  pr,  tmp);
+    *residual = nrm2(n, tmp);
+    if (*residual < tolerance)
+    {
+        scal(n, (ValueType)1.0/nrm1(n,pr), pr);
+        return true;
+    }
+    else
+    {
+        if (iter< max_iter)
+        {
+            std::swap(pr, tmp);
+        }
+        else
+        {
+           scal(n, (ValueType)1.0/nrm1(n,pr), pr);
+        }
+        return false;
+    }
+}
+
+template <typename IndexType, typename ValueType>
+int pagerankSolver(IndexType n, IndexType e, IndexType const *cscPtr, IndexType const *cscInd, ValueType *cscVal,
+             IndexType *prsVtx, ValueType *prsVal, IndexType prsLen, bool has_personalization,
+             ValueType alpha, ValueType *a, bool has_guess, float tolerance, int max_iter,
+             ValueType * &pagerank_vector, ValueType * &residual) {
+  int max_it, i = 0 ;
+  float tol;
+  bool converged = false;
+  ValueType randomProbability = static_cast<ValueType>( 1.0/n);
+  ValueType *b=0, *tmp=0;
+  void* cub_d_temp_storage = NULL;
+  size_t cub_temp_storage_bytes = 0;
+
+  if (max_iter > 0)
+      max_it = max_iter;
+  else
+      max_it =  500;
+
+  if (tolerance == 0.0f)
+      tol =  1.0E-6f;
+  else if (tolerance < 1.0f && tolerance > 0.0f)
+      tol = tolerance;
+  else
+      return -1;
+
+  if (alpha <= 0.0f || alpha >= 1.0f)
+      return -1;
+
+  cudaStream_t stream{nullptr};
+
+  ALLOC_TRY((void**)&b, sizeof(ValueType) * n, stream);
```

Comment:
We should start replacing ALLOC_TRY with rmm::device_vector or rmm::device_buffer.

Reply:
Made this change where it could be made.